File: LeetcodeProblems/blind-75/dp/palindromic-substrings/solution.py
# Palindromic Substrings - LeetCode #647

class Solution:
    def countSubstrings(self, s: str) -> int:
        count = 0
        
        for i in range(len(s)):
            # Odd length
            count += self.expand(s, i, i)
            # Even length
            count += self.expand(s, i, i + 1)
        
        return count
    
    def expand(self, s: str, left: int, right: int) -> int:
        local_count = 0
        while left >= 0 and right < len(s) and s[left] == s[right]:
            local_count += 1
            left -= 1
            right += 1
        return local_count

File: LeetcodeProblems/blind-75/dp/palindromic-substrings/solution.js
// Palindromic Substrings - LeetCode #647

/**
 * @param {string} s
 * @return {number}
 */
var countSubstrings = function(s) {
    let count = 0;
    
    for (let i = 0; i < s.length; i++) {
        // Odd length
        count += expand(s, i, i);
        // Even length
        count += expand(s, i, i + 1);
    }
    
    return count;
};

function expand(s, left, right) {
    let count = 0;
    while (left >= 0 && right < s.length && s[left] === s[right]) {
        count++;
        left--;
        right++;
    }
    return count;
}
